DR-2031 Handle field delimiter and quote characters for CSV

tlangs opened the pull request:
It looks like Azure Synapse doesn't support null characters and doesn't care about quoted newlines, so we won't be implementing those. Instead, we'll return a 400 and say that these features aren't supported on Azure.

This PR implements the field delimiter and quote character for Azure Synapse.

Commit:
handle field delimiter and quote characters for CSV

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/AzureSynapsePdaoConnectedTest.java:
```diff
@@ -155,6 +154,30 @@ public void testSynapseQueryCSV() throws Exception {
+    assertThat(
+        "The text columns should be properly quoted",
+        textCols,
+        equalTo(Arrays.asList("Dao!", "Jones!")));
```
```suggestion
        equalTo(Lists.of("Dao!", "Jones!")));
```

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -196,11 +203,16 @@ public int createParquetFiles(
     sqlCreateTableTemplate.add("isCSV", isCSV);
     if (isCSV) {
       sqlCreateTableTemplate.add("parserVersion", PARSER_VERSION);
-      if (csvSkipLeadingRows.isEmpty()) {
+      if (csvSkipLeadingRows == null) {
         throw new InvalidJobParameterException(
             "For CSV ingests, 'csvSkipLeadingRows' must be defined.");
       }
-      sqlCreateTableTemplate.add("firstRow", csvSkipLeadingRows.get());
+      sqlCreateTableTemplate.add("firstRow", csvSkipLeadingRows);
```
If you don't mind throwing `NullPointerException`, you can replace with
```
      sqlCreateTableTemplate.add("firstRow", 
        Objects.requireNonNull(csvSkipLeadingRows, "For CSV ingests, 'csvSkipLeadingRows' must be defined."));
```

If you don't want to throw an NPE, you can replace with
```
      sqlCreateTableTemplate.add("firstRow", 
        Objects.requireNonNullElseGet(csvSkipLeadingRows, 
            () -> { throw new InvalidJobParameterException("For CSV ingests, 'csvSkipLeadingRows' must be defined." })));
```

pshapiro4broad reviewed (approved):
Looks ok, just a few minor things

Commit:
Move validation to request processing instead of flight

tlangs commented:
@pshapiro4broad I've moved the ingest request validation out of the Pdao and into the controller layer, since its much better to get an invalid parameter response quickly rather than spinning up a job and making users wait for that to error out. I stuck the validation code in `IngestRequestValidator`, but I'm unsure if that's exactly the right place for it, since `IngestRequestValidator` is automatically wired in and here I'm calling a method that I added on it manually. Thoughts?

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -183,7 +187,9 @@ public int createParquetFiles(
       String controlFileDataSourceName,
       String ingestTableName,
-      Optional<Integer> csvSkipLeadingRows)
+      Integer csvSkipLeadingRows,
```
now that you know it can't be null, you can use `int`
```suggestion
      int csvSkipLeadingRows,
```

tlangs replied:
Unfortunately, we still pass this parameter in for JSON ingests, and it can be null for that.

pshapiro4broad replied:
ah, of course :)

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/IngestRequestValidator.java:
```diff
@@ -22,6 +35,36 @@ private void validateTableName(String name, Errors errors) {
+    if (platform.isAzure()) {
+      return validateAzureIngestParams(ingestRequestModel);
+    }
+    return Collections.emptyList();
```
```suggestion
    return List.of();
```

tlangs replied:
Any reason for one over the other?

pshapiro4broad replied:
`List.of()` is a more general form since it supports zero, one or many args. It's also nice in that it's in `List` instead of `Collections`, so it's more tied to the type we're using. So I've been using `List.of()` (and `Map.of()`, `Set.of()` etc) wherever I used to use the collection methods.

But that's it, it's not more efficient or better in any measurable way. So either way is really fine.

tlangs replied:
This actually becomes a non-issue since I'm changing the way errors are handled!

pshapiro4broad reviewed (approved):
Yeah, I'm not sure if `IngestRequestValidator` is the correct place since that's a spring validator which has a special role. I think I'd put it in `DatasetsApiController` instead since that's the only place it's used, and that context already injects `DatasetService`.

If you wanted to, you could create a unit test for the validation code (using a mock of `DatasetService`).

tlangs commented:
Good suggestion. I'll move it to `DatasetService`

pshapiro4broad commented:
> Good suggestion. I'll move it to DatasetService

Actually I was thinking `DatasetsApiController` since the validation deals with the API layer model class.

Commit:
Fixed tests and moved validation

snf2ye commented on src/main/java/bio/terra/app/controller/DatasetsApiController.java:
```diff
@@ -328,4 +333,49 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
+  }
+
+  private void validateAzureIngestParams(IngestRequestModel ingestRequest) {
+    List<String> errors = new ArrayList<>();
```
I vaguely remember discussing throwing an error if the user tries to set the "csv_allow_quoted_newlines"  or "csv_null_marker" for Azure ingest since we don't support them for Azure. Could we include this here?

tlangs replied:
Oh I talked about this with Muscles. We provide defaults for those two fields, so it would be impossible to tell if a user defined a value that is the same as the default, or if it was just the default that we provide. Since we can't determine whether its a user action or not, we decided that we should just document that the fields will be ignored for Azure.

snf2ye reviewed (commented):
Otherwise LGTM! Thank you!
